yarn-project/world-state/src/synchronizer/server_world_state_synchronizer.ts

```diff
@@ -138,46 +170,56 @@ export class ServerWorldStateSynchronizer implements WorldStateSynchronizer {
    * @returns A promise that resolves with the block number the world state was synced to
    */
-  public async syncImmediate(minBlockNumber?: number): Promise<number> {
+  public async syncImmediate(targetBlockNumber?: number): Promise<number> {
     if (this.currentState !== WorldStateRunningState.RUNNING) {
       throw new Error(`World State is not running, unable to perform sync`);
     }
-    // If we have been given a block number to sync to and we have reached that number
-    // then return.
-    if (minBlockNumber !== undefined && minBlockNumber <= this.currentL2BlockNum) {
+    // If we have been given a block number to sync to and we have reached that number then return.
+    if (targetBlockNumber !== undefined && targetBlockNumber <= this.currentL2BlockNum) {
       return this.currentL2BlockNum;
     }
-    const blockToSyncTo = minBlockNumber === undefined ? 'latest' : `${minBlockNumber}`;
-    this.log.debug(`World State at block ${this.currentL2BlockNum}, told to sync to block ${blockToSyncTo}...`);
-    // ensure any outstanding block updates are completed first.
+    this.log.debug(`World State at ${this.currentL2BlockNum} told to sync to ${targetBlockNumber ?? 'latest'}`);
+    // ensure any outstanding block updates are completed first
     await this.jobQueue.syncPoint();
+
     while (true) {
       // Check the block number again
-      if (minBlockNumber !== undefined && minBlockNumber <= this.currentL2BlockNum) {
+      if (targetBlockNumber !== undefined && targetBlockNumber <= this.currentL2BlockNum) {
         return this.currentL2BlockNum;
       }
       // Poll for more blocks
-      const numBlocks = await this.l2BlockDownloader.pollImmediate();
+      const numBlocks = await this.l2BlockDownloader.pollImmediate(targetBlockNumber);
       this.log.debug(`Block download immediate poll yielded ${numBlocks} blocks`);
       if (numBlocks) {
         // More blocks were received, process them and go round again
         await this.jobQueue.put(() => this.collectAndProcessBlocks());
         continue;
       }
       // No blocks are available, if we have been given a block number then we can't achieve it
-      if (minBlockNumber !== undefined) {
+      if (targetBlockNumber !== undefined) {
         throw new Error(
-          `Unable to sync to block number ${minBlockNumber}, currently synced to block ${this.currentL2BlockNum}`,
+          `Unable to sync to block number ${targetBlockNumber}, currently synced to block ${this.currentL2BlockNum}`,
         );
       }
       return this.currentL2BlockNum;
     }
   }
 
+  public async syncImmediateAndFork(
+    targetBlockNumber: number,
+    forkIncludeUncommitted: boolean,
+  ): Promise<MerkleTreeOperationsFacade> {
+    await this.pause();
+    await this.syncImmediate(targetBlockNumber);
+    const fork = await this.getFork(forkIncludeUncommitted);
+    this.resume();
```

Comment:
I'm wondering if this should be a `try .. finally`?
```suggestion
    await this.pause();
    try {
      await this.syncImmediate(targetBlockNumber);
      return await this.getFork(forkIncludeUncommitted);
    } finally {
      this.resume();
    }
```